### PR TITLE
Update individual trigger resource

### PR DESCRIPTION
**Pull request by shivacharanms**
This PR contains 

- telemetry service handler to update triggers by creating tasks for each plugin
- Base code in grf plugin to get the saved inventory of servers and forward the patch request
- API service update handler change to send the request body part of rpc call

### CODE REVIEW

File: svc-api/handle/telemetry.go [hunk 1]
```diff
@@ -246,7 +247,6 @@ func (a *TelemetryRPCs) GetMetricReportDefinition(ctx iris.Context) {
 func (a *TelemetryRPCs) GetMetricReport(ctx iris.Context) {
 	req := telemetryproto.TelemetryRequest{
 		SessionToken: ctx.Request().Header.Get("X-Auth-Token"),
-		URL:          ctx.Request().RequestURI,
```
Comment (bharath-b-hpe) [inline svc-api/handle/telemetry.go:249]: This was added by Irappa for storing metric data in DB, please check if it got removed during pull.

File: svc-telemetry/tcommon/common.go [hunk 2]
```diff
@@ -137,15 +138,18 @@ func ContactPlugin(req PluginContactRequest, errorMessage string) ([]byte, strin
 	var resp ResponseStatus
 	var response *http.Response
 	var err error
+	log.Info("INSIDE contact client")
```
Comment (bharath-b-hpe) [inline svc-telemetry/tcommon/common.go:141]: Please remove debug logs in this file, if not required.

**Review by jeevan-kamkar — APPROVED** [review]

File: plugin-redfish/rfphandler/telemetry_test.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+	deviceHost := "localhost"
+	devicePort := "1234"
+	ts := startTestServer(mockUpdateTrigger)
+	// Start the server.
+	ts.StartTLS()
+	defer ts.Close()
+	mockApp := iris.New()
+	redfishRoutes := mockApp.Party("/ODIM/v1")
+
+	redfishRoutes.Post("/TelemetryService/Triggers/sample", SimpleUpdate)
```
Comment (irappa-pattar) [inline plugin-redfish/rfphandler/telemetry_test.go:44]: SimpleUpdate? please update the proper method and its patch request.

File: plugin-redfish/rfphandler/telemetry_test.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	test := httptest.New(t, mockApp)
+	attributes := map[string]interface{}{"EventTriggers": []string{"Alert"}}
+	attributeByte, _ := json.Marshal(attributes)
+	requestBody := map[string]interface{}{
+		"ManagerAddress": fmt.Sprintf("%s:%s", deviceHost, devicePort),
+		"UserName":       "admin",
+		"Password":       []byte("P@$$w0rd"),
+		"PostBody":       attributeByte,
+	}
+	test.POST("/ODIM/v1/TelemetryService/Triggers/sample").WithJSON(requestBody).Expect().Status(http.StatusOK)
```
Comment (irappa-pattar) [inline plugin-redfish/rfphandler/telemetry_test.go:55]: PATCH request

File: svc-telemetry/telemetry/telemetry.go [hunk 5]
```diff
@@ -373,17 +377,231 @@ func (e *ExternalInterface) GetTrigger(req *teleproto.TelemetryRequest) response
...
+		return resp
+	}
+
+	pluginList, err := tmodel.GetAllKeysFromTable("Plugin", common.OnDisk)
+	if err != nil {
+		errMsg := "Request parameters validaton failed: " + err.Error()
+		log.Error(errMsg)
+		return common.GeneralError(http.StatusInternalServerError, response.InternalError, errMsg, nil, nil)
+	}
+	log.Info("Count of plugin list")
```
Comment (irappa-pattar) [inline svc-telemetry/telemetry/telemetry.go:422]: remove unwanted debug logs

**Review by irappa-pattar — CHANGES_REQUESTED** [review]
please address the comments and does the patch request exposed in the router?

File: plugin-redfish/rfphandler/telemetry_test.go (hunk 3 above)
Comment (Bharath-KKB) [reply plugin-redfish/rfphandler/telemetry_test.go:44]: /TelemetryService/Triggers/sample is correct URI ?

File: plugin-redfish/rfphandler/telemetry_test.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	"github.com/ODIM-Project/ODIM/plugin-redfish/config"
+	"github.com/ODIM-Project/ODIM/plugin-redfish/rfpresponse"
+	iris "github.com/kataras/iris/v12"
+	"github.com/kataras/iris/v12/httptest"
+	"net/http"
+	"testing"
+)
+
+func mockUpdateTrigger(username, password, url string, w http.ResponseWriter) {
+	w.WriteHeader(http.StatusOK)
```
Comment (Bharath-KKB) [inline plugin-redfish/rfphandler/telemetry_test.go:29]: what do we achieve with this ?

File: plugin-redfish/rfphandler/update_test.go [hunk 7]
```diff
@@ -1,3 +1,17 @@
+//(C) Copyright [2020] Hewlett Packard Enterprise Development LP
```
Comment (Bharath-KKB) [inline plugin-redfish/rfphandler/update_test.go:1]: this looks like simplupdate test. What has this to do with current story ?

File: plugin-redfish/rfphandler/telemetry_test.go (hunk 3 above)
Comment (Bharath-KKB) [reply plugin-redfish/rfphandler/telemetry_test.go:44]: Also this looks like it is testing POST. Is update done using POST ?

File: plugin-redfish/rfphandler/telemetry_test.go [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+func TestUpdateTrigger(t *testing.T) {
+	config.SetUpMockConfig(t)
+
+	deviceHost := "localhost"
+	devicePort := "1234"
+	ts := startTestServer(mockUpdateTrigger)
```
Comment (Bharath-KKB) [inline plugin-redfish/rfphandler/telemetry_test.go:37]: what does this call do ?

**Review by Bharath-KKB — CHANGES_REQUESTED** [review]
address comments

File: svc-api/handle/handlers.go [hunk 9]
```diff
@@ -152,7 +152,7 @@ func getService(microServices []string, uuid string) models.ServiceRoot {
 					serviceRoot.UpdateService = &models.Service{OdataID: servicePath}
 				}
 			}
-		case "TelemetryService":
+		case "Telemetry":
```
Comment (Bharath-KKB) [inline svc-api/handle/handlers.go:155]: Telemetry is a redfish service. I think we should rename it as TelemetryService.

**Review by Bharath-KKB — CHANGES_REQUESTED** [review]
address comments

File: svc-api/handle/telemetry.go (hunk 1 above)
Comment (shivacharanms) [reply svc-api/handle/telemetry.go]: addressed

File: svc-telemetry/tcommon/common.go (hunk 2 above)
Comment (shivacharanms) [reply svc-telemetry/tcommon/common.go]: addressed

File: plugin-redfish/rfphandler/telemetry_test.go (hunk 3 above)
Comment (shivacharanms) [reply plugin-redfish/rfphandler/telemetry_test.go]: addressed

File: plugin-redfish/rfphandler/telemetry_test.go (hunk 4 above)
Comment (shivacharanms) [reply plugin-redfish/rfphandler/telemetry_test.go]: addressed

File: plugin-redfish/rfphandler/telemetry_test.go (hunk 6 above)
Comment (shivacharanms) [reply plugin-redfish/rfphandler/telemetry_test.go]: changed it to handler response code coming from handler

File: plugin-redfish/rfphandler/update_test.go (hunk 7 above)
Comment (shivacharanms) [reply plugin-redfish/rfphandler/update_test.go:1]: License was missing

File: plugin-redfish/rfphandler/telemetry_test.go (hunk 8 above)
Comment (shivacharanms) [reply plugin-redfish/rfphandler/telemetry_test.go]: For all UT under plugin, we have a test server running which handles our server calls https://github.com/ODIM-Project/ODIM/blob/development/plugin-redfish/rfphandler/biosSettings_test.go#L155

File: svc-api/handle/handlers.go (hunk 9 above)
Comment (shivacharanms) [reply svc-api/handle/handlers.go]: Yes, i have edited it back to TelemetryService

File: plugin-redfish/rfphandler/telemetry_test.go [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/ODIM-Project/ODIM/plugin-redfish/rfpresponse"
+	iris "github.com/kataras/iris/v12"
+	"github.com/kataras/iris/v12/httptest"
+	"io/ioutil"
+	"net/http"
+	"testing"
+)
+
+func mockUpdateTriggerHandler(username, password, url string, w http.ResponseWriter) {
+	resp, err := mockChangeBiosSettings(username, url)
```
Comment (Bharath-KKB) [inline plugin-redfish/rfphandler/telemetry_test.go:31]: What is mockChangeBiosSettings doing here ? It is not being used anywhere as well.

File: plugin-redfish/rfphandler/update_test.go (hunk 7 above)
Comment (Bharath-KKB) [reply plugin-redfish/rfphandler/update_test.go:1]: That then should have been a seperate story and PR

File: svc-api/handle/telemetry_test.go [hunk 11]
```diff
@@ -42,6 +42,29 @@ func testTelemetryService(req teleproto.TelemetryRequest) (*teleproto.TelemetryR
 	return response, nil
 }
 
+func mockUpdateTriger(req teleproto.TelemetryRequest) (*teleproto.TelemetryResponse, error) {
```
Comment (Bharath-KKB) [inline svc-api/handle/telemetry_test.go:45]: thanks for actioning this. I suggested using anon funcs to Jeevan. That should make these more manageable.

File: svc-api/handle/telemetry_test.go [hunk 12]
```diff
@@ -206,18 +229,15 @@ func TestGetTrigger(t *testing.T) {
 
 func TestUpdateTrigger(t *testing.T) {
 	var a TelemetryRPCs
-	a.UpdateTriggerRPC = testTelemetryService
+	a.UpdateTriggerRPC = mockUpdateTriger
```
Comment (Bharath-KKB) [inline svc-api/handle/telemetry_test.go:232]: Trigger

**Review by Bharath-KKB — CHANGES_REQUESTED** [review]
A few more comments

File: plugin-redfish/rfphandler/telemetry.go [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+			notFoundCount++
+		default:
+			internalErrorCount++
+		}
+
+	}
+	if successCount > 0 {
+		ctx.StatusCode(http.StatusOK)
+		return
+	}
```
Comment (bharath-b-hpe) [inline plugin-redfish/rfphandler/telemetry.go:74]: Please handle scenario when successCount is 0 and internalErrorCount is greater than 0.

File: svc-api/handle/telemetry.go [hunk 14]
```diff
@@ -303,8 +304,30 @@ func (a *TelemetryRPCs) GetTrigger(ctx iris.Context) {
...
+	err := ctx.ReadJSON(&reqBody)
+	if err != nil {
+		errorMessage := "error while trying to get JSON body from the account update request body: " + err.Error()
+		log.Error(errorMessage)
+		response := common.GeneralError(http.StatusBadRequest, response.MalformedJSON, errorMessage, nil, nil)
+		ctx.StatusCode(http.StatusBadRequest) // TODO: add error headers
+		ctx.JSON(&response.Body)
+		return
+	}
+	request, err := json.Marshal(reqBody)
```
Comment (bharath-b-hpe) [inline svc-api/handle/telemetry.go:318]: Please replace ctx.ReadJSON and json.Marshal with single call to ctx.GetBody() ([]byte, error) to get the payload in byte array.

File: svc-telemetry/telemetry/telemetry.go [hunk 15]
```diff
@@ -373,17 +377,226 @@ func (e *ExternalInterface) GetTrigger(req *teleproto.TelemetryRequest) response
...
+	json.Unmarshal(requestBody, &updateRequest)
+	if len(updateRequest) <= 0 {
+		return true
+	}
+	return false
+}
+
+func formTargetList(keys []string) []tmodel.Plugin {
+	var plugins []tmodel.Plugin
+	for _, key := range keys {
```
Comment (bharath-b-hpe) [inline svc-telemetry/telemetry/telemetry.go:593]: Please check if plugin is of Compute type.

File: svc-api/handle/telemetry_test.go (hunk 12 above)
Comment (shivacharanms) [reply svc-api/handle/telemetry_test.go]: Updated

File: plugin-redfish/rfphandler/telemetry_test.go (hunk 10 above)
Comment (shivacharanms) [reply plugin-redfish/rfphandler/telemetry_test.go]: corrected

File: svc-api/handle/telemetry.go (hunk 14 above)
Comment (shivacharanms) [reply svc-api/handle/telemetry.go]: Done

**Review by bharath-b-hpe — APPROVED** [review]

**Review by Bharath-KKB — APPROVED** [review]

**Review by irappa-pattar — APPROVED** [review]

**Pull request merged**